modules/rap/bundles/runtime/org.eclipse.rcptt.runtime.ui.rap/src/org/eclipse/rcptt/internal/runtime/ui/rap/AUTControlsHierarchyService.java

```diff
@@ -0,0 +1,178 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2016 Xored Software Inc and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Xored Software Inc - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+package org.eclipse.rcptt.internal.runtime.ui.rap;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.rcptt.core.ecl.core.model.AUTControlsHierarchy;
+import org.eclipse.rcptt.ecl.core.Command;
+import org.eclipse.rcptt.ecl.runtime.ICommandService;
+import org.eclipse.rcptt.ecl.runtime.IProcess;
+import org.eclipse.rcptt.tesla.core.protocol.ProtocolFactory;
+import org.eclipse.rcptt.tesla.core.protocol.UIHierarchyResponse;
+import org.eclipse.rcptt.tesla.core.protocol.raw.Element;
+import org.eclipse.rcptt.tesla.internal.ui.player.SWTUIElement;
+import org.eclipse.rcptt.tesla.recording.aspects.swt.rap.SWTEventManager;
+import org.eclipse.rcptt.tesla.ui.RWTUtils;
+import org.eclipse.swt.widgets.Display;
+
+public class AUTControlsHierarchyService implements ICommandService {
+
+	private IProcess context = null;
+
+	private Display display = null;
+
+	public IStatus service(Command command, IProcess context)
+			throws InterruptedException, CoreException {
+
+		this.context = context;
+
+		AUTControlsHierarchy autControlsHierarchy = (AUTControlsHierarchy) command;
+
+		switch (autControlsHierarchy.getState()) {
+
+		case HIGHLIGHT_WIDGET:
+			return highlightWidget(autControlsHierarchy);
+
+		case UPDATE_ASSERT_WINDOW:
+			return updateAssertionPanelWindow(autControlsHierarchy);
+
+		case GET_ELEMENT:
+			return getElement(autControlsHierarchy);
+
+		case GET_PARENT:
+			return getParent(autControlsHierarchy);
+
+		default:
+			return Status.OK_STATUS;
+		}
+	}
+
+	private void write(Object object) throws CoreException {
+		context.getOutput().write(object);
+	}
+
+	private Display getDisplay() {
+		if (null == display) {
```

Comment:
In case of RAP display could be different in different sessions @nomit007 please check. Do not cache it please.